pkg/gpu/ebpf/c/gpu.c

```diff
@@ -0,0 +1,139 @@
+    launch_data.header.ktime_ns = bpf_ktime_get_ns();
+    launch_data.header.stream_id = (uint64_t)stream;
+    launch_data.header.type = cuda_kernel_launch;
+    launch_data.kernel_addr = (uint64_t)func;
+    launch_data.shared_mem_size = shared_mem;
+
+    log_debug("cudaLaunchKernel: EMIT[1/2] pid_tgid=%llu, ts=%llu", launch_data.header.pid_tgid, launch_data.header.ktime_ns);
+    log_debug("cudaLaunchKernel: EMIT[2/2] kernel_addr=0x%llx, shared_mem=%llu, stream_id=%llu", launch_data.kernel_addr, launch_data.shared_mem_size, launch_data.header.stream_id);
+
+    bpf_ringbuf_output(&cuda_events, &launch_data, sizeof(launch_data), 0);
+
+    return 0;
+}
+
+SEC("uprobe/cudaMalloc")
+int BPF_UPROBE(uprobe__cudaMalloc, void **devPtr, size_t size) {
+    __u64 pid_tgid = bpf_get_current_pid_tgid();
+    cuda_alloc_request_args_t args = { .devPtr = devPtr, .size = size };
+
+    log_debug("cudaMalloc: pid=%llu, devPtr=%llx, size=%lu", pid_tgid, (__u64)devPtr, size);
+    bpf_map_update_elem(&cuda_alloc_cache, &pid_tgid, &args, BPF_ANY);
+
+    return 0;
+}
+
+SEC("uretprobe/cudaMalloc")
+int BPF_URETPROBE(uretprobe__cudaMalloc) {
+    __u64 pid_tgid = bpf_get_current_pid_tgid();
+    cuda_alloc_request_args_t *args;
+    cuda_memory_event_t mem_data;
+
+    log_debug("cudaMalloc[ret]: pid=%llx\n", pid_tgid);
+
+    args = bpf_map_lookup_elem(&cuda_alloc_cache, &pid_tgid);
+    if (!args) {
+        log_debug("cudaMalloc[ret]: failed to find cudaMalloc request");
+        return 0;
+    }
+
+    __builtin_memset(&mem_data, 0, sizeof(mem_data));
+
+    mem_data.header.pid_tgid = bpf_get_current_pid_tgid();
+    mem_data.header.stream_id = (uint64_t)0;
+    mem_data.header.type = cuda_memory_event;
+    mem_data.header.ktime_ns = bpf_ktime_get_ns();
+    mem_data.type = cudaMalloc;
+    mem_data.size = args->size;
+
+    if (bpf_probe_read_user(&mem_data.addr, sizeof(void *), args->devPtr)) {
+        log_debug("cudaMalloc[ret]: failed to read devPtr from cudaMalloc at 0x%llx", (__u64)args->devPtr);
+        return 0;
+    }
+
+    log_debug("cudaMalloc[ret]: EMIT size=%llu, addr=0x%llx, ts=%llu", mem_data.size, (__u64)mem_data.addr, mem_data.header.ktime_ns);
+
+    bpf_ringbuf_output(&cuda_events, &mem_data, sizeof(mem_data), 0);
+
+    return 0;
+}
+
+SEC("uprobe/cudaFree")
+int BPF_UPROBE(uprobe__cudaFree, void *mem) {
+    cuda_memory_event_t mem_data;
+
+    __builtin_memset(&mem_data, 0, sizeof(mem_data));
+
+    mem_data.header.pid_tgid = bpf_get_current_pid_tgid();
+    mem_data.header.stream_id = (uint64_t)0;
+    mem_data.header.type = cuda_memory_event;
+    mem_data.header.ktime_ns = bpf_ktime_get_ns();
+    mem_data.size = 0;
+    mem_data.addr = (uint64_t)mem;
+    mem_data.type = cudaFree;
+
+    bpf_ringbuf_output(&cuda_events, &mem_data, sizeof(mem_data), 0);
+
+    return 0;
+}
+
+SEC("uprobe/cudaStreamSynchronize")
+int BPF_UPROBE(uprobe__cudaStreamSynchronize, __u64 stream) {
+    // TODO: Send this on return, not on entry
```

Comment:
why though? it can fail?

Reply:
No, but the timing should probably be recorded in the return. I left it here for simplicity and then forgot, I will change it.